urllib3/connectionpool.py

```diff
@@ -628,6 +627,12 @@ def urlopen(self, method, url, body=None, headers=None, retries=None,
             err = e
 
         finally:
+            if not clean_exit:
+                # We hit some kind of exception, handled or otherwise. We need
+                # to throw the connection away unless explicitly told not to.
+                conn = conn and conn.close()
```

Comment:
Exactly right. This pattern was present in the other blocks of code that I consolidated here.